jekyll/_posts/2021-07-21-Nim-Efficienct-Expressive-Elecgant-Benchmarking.md

```diff
@@ -0,0 +1,369 @@
+        `cptr0id`[] = `cptr0id`[] or cBITMSK[`cullaid` and 7]
+        `cullaid` += `bpintid`
+    csstmnt.add nnkOfBranch.newTree(
+      newLit(n),
+      ofbrstmnts
+    )
+  for n in 0x40'u8 .. 255'u8: # fill in defaults for remaining possibilities
+    csstmnt.add nnkOfBranch.newTree(
+      newLit(n),
+      nnkStmtList.newTree(
+        nnkBreakStmt.newTree(
+          newEmptyNode()
+        )
+      )
+    )
+  result.add quote do:
+    let `endalmtid` = `cmpstsalmtid` - `strt7id`
+    var `cullaid` = `ca` + `strt0id`
+    `csstmnt`
+#  echo csstmnt[9].astGenRepr # see AST for a given case
+#  echo csstmnt[9].toStrLit # see code for a given case
+#  echo result.toStrLit # see entire produced code at compile time
+
+proc benchSoE(): iterator(): int {.closure.} =
+  var cmpsts = newSeq[byte](16384)
+  let cmpstsa = cast[int](cmpsts[0].addr)
+  for _ in 0 ..< cLOOPS:
+    for i in 0 .. 254: # cull to square root of limit
+      if (cmpsts[i shr 3] and cBITMSK[i and 7]) == 0'u8: # if prime -> cull its composites
+        let bp = i +% i +% 3
+        let swi = (i +% i) *% (i +% 3) +% 3
+        unrollLoops(cmpstsa, 16384, swi, bp)
+  return iterator(): int {.closure.} =
+    yield 2 # the only even prime
+    for i in 0 .. 131071: # separate iteration over results
+      if (cmpsts[i shr 3] and cBITMSK[i and 7]) == 0'u8:
+        yield i +% i +% 3
+```
+Running the code with the above substitution in the same non-checking release mode as before results in a run time that is about 2.5 times faster than the previous version and over ten times faster than the naive implementation, meaning that the culling operation time is this ratio faster per operation.  For a modern CPU, one can calculate that it uses just over one CPU clock cycle per cull operation (198273 culling operations per loop).
+
+Now, this speed can't quite be maintained for a real application with larger ranges and thus larger base prime values as there will come a point where the overheads of the loop unrolling overcomes its benefit, but that is for quite large ranges of about a hundred million; as well, the loss above that point increases slowly since the base prime values only grow as the square root of the entire sieved range and also the base primes density becomes more sparse with increasing range such that using a page segmented implementation to count the primes up to ten's of billions only increases the time per culling operation to about two CPU clock cycles each, meaning that the average is still much less than pure "bit-twiddling".
+
+## Analysis as compared to other languages
+
+So in order to be able to assert that Nim is "Efficient, Effective, and Elegant", one needs to compare it to something; since Nim's design goals are in that order, languages that aren't as efficient will be rejected from the outset.  This means the following:
+
+1. Languages such as Python and R are not comparable because their default implementations are interpreted and thus much slower.
```

Comment:
I would put it as "are not comparable because they use dynamic typing and the evidence is now that static typing is more effective" (<insert quote to the study here>)

Reply:
I can add that in this small section about interpreted and/or dynamic languages;  As mentioned, Julia is a strange beast in that it is dynamically typed but can be statically guided, and can produce some very fast code as long as one stays away from functional type programming as Julia doesn't have a type signature for functions and needs to determine them at runtime, making closure calls extremely slow (ten times slower).  Along with it's "non-symetrical `end`'s, it isn't really a language I enjoy, but it can produce very fast code - sometimes.  Libraries are pretty spotty as well, as contributors will do weird things with dynamic typing just because they can.

Reply:
It doesn't matter how fast Python/R/Julia can be made to perform, they lack static typing. Give me a Python that is 10x faster than Nim and I still don't want to use Python. Just like I don't use LuaJIT. Dynamic typing (for me personally anyway) is a productivity killer and misfeature like no other.

Reply:
You are "preaching to the choir" here about static typing; all of my favorite languages are static.  Someone once said that dynamically typed languages aren't typed at all, as everything just is a be-everything object that is tagged with what it need to represent in each of its guises.

However, I feel an obligation to try a lot of languages just to see what can be learned from them.

Yes, I lost a lot of time fighting Julia's "guided type" dynamic type system, but saw that it could do certain things quite well, other things not-so-much.  Then I left it because it is particularly bad/slow at functional types of algorithms.  It seems to have an appeal to those who use dynamically typed languages such as Python, Wolfram language (Mathematica), Matlab/Octave, and R because they are used to the limitations of dynamic typing, and with Julia they can get quite a boost in speed without having to learn C/C++.  Theoretically, Nim could appeal to them too, but it isn't sourced from academia, so its a hard sell, and then there is Julia's huge mathematical libraries that seem to match the capabilities of all of the languages that are slow.

Now that we have Arc/Orc, it has kind of caused me to have a distaste for conventional garbage collectors as well, so while Haskell and F# have a place in my heart, it's not what it once was.

Reply:
> You are "preaching to the choir" here about static typing;

I'm really bad at getting the point across. I don't mean to convince you. There is a common line of thinking "out there" that is much like "oh Python would be perfect if only it was faster" and I don't like to strengthen these ideas. But a sentence like "Languages such as Python and R are not comparable because their default implementations are interpreted and thus much slower" does do that and so I mind the sentence. If you want to compare to Python/R/whatever focus on differences we care about, speed is just an "implementation detail".

Reply:
@Araq:

> There is a common line of thinking "out there" that is much like "oh Python would be perfect if only it was faster" and I don't like to strengthen these ideas. But a sentence like "Languages such as Python and R are not comparable because their default implementations are interpreted and thus much slower" does do that and so I mind the sentence. If you want to compare to Python/R/whatever focus on differences we care about, speed is just an "implementation detail".

Ah, now I see your point and your objection.  The reason that the comparisons in this section are mostly about speed is that this is an article on benchmarks and therefore mostly focused on code performance/speed/"Efficiency" while still maintaining "Effectiveness" and "Elegance".

Therefore, by the phrase "Languages such as Python and R are not comparable because their default implementations are interpreted and thus much slower", I meant **in the context of performance benchmarks and "efficiency"** and also quickly wrote them off as they fill a different use case than Nim.

I will edit that section to make it more clear on why they are being excluded from direct comparison to Nim, not only for reasons of speed (Julia can be fast, as explained in the JIT languages section, in which I also mention that one of its major disadvantages is being dynamically typed) but because they can't be used for many of the programming tasks that Nim can, and one of the reasons for that is dynamic typing.  Something like the following:

"""
1. Languages such as Python and R are not comparable in the context of performance/efficiency because their default implementations are interpreted and thus much slower.  As well, almost all languages in this same category are dynamically typed (which contributes to their lack of competitive performance), which make them ineffective if applied to the many of the kinds of programming tasks to which NIm can be applied.  Among the host of other languages in this category are all of the Lisp-based languages such as Scheme/Racket/Common Lisp and also Clojure, which is both a dynamically typed language and a Just In Time compiled language that also fits into the next section.
"""

So how should I proceed on this PR?  I need to properly rename the file and edit the above section if you agree, as well make other changes as suggested below.  Should I close/cancel this PR and just open a new one, which would seem to be the easiest thing to do in the interests of renaming the file?

Reply:
You can simple keep pushing into this branch and it will be updated. I think. "close and reopen" is fine with me too, whatever works for you.

Reply:
> which would seem to be the easiest thing to do in the interests of renaming the file?

@GordonBGood you can run something like:
```
git clone https://github.com/GordonBGood/website gbgwebsite
cd gbgwebsite/jekyll/_posts
git mv 2021-07-21-Nim-Efficienct-Expressive-Elecgant-Benchmarking.md 2021-07-21-nim-efficient-expressive-elegant-benchmarking.md
git commit
git push
```

Alternatively, somebody with write access to the `nim-lang/website` repo could run that, assuming that the "Allow edits by maintainers" checkbox is ticked in the PR. 

Note that this PR corresponds to the `master` branch of the fork.

Reply:
I think it's easier, at least for me, to just close this PR, make the changes on my fork, and post a new PR.  Closing now.